tests/preprocessing/test_imputation.py

```diff
@@ -64,158 +71,237 @@ def test_median_impute_throws_error_non_numerical(impute_adata):
+    assert np.sum([adata_imputed.X == 1011]) == 1
+    assert np.sum([adata_imputed.X == "REPLACED"]) == 1
 
 
 def test_warning(impute_num_adata):
     warning_results = _warn_imputation_threshold(impute_num_adata, threshold=20, var_names=None)
     assert warning_results == {"col1": 25, "col3": 50}
+
+
+# ----------------------------------------------------------------------------------------------------------------------
+# Below: tests to check if _base_check_imputation works as intended (test-inception)
+# ----------------------------------------------------------------------------------------------------------------------
+def test_base_check_imputation_incompatible_shapes(impute_num_adata):
+    adata_imputed = knn_impute(impute_num_adata, copy=True)
+    assert not _base_check_imputation(impute_num_adata, adata_imputed[1:, :]) and not _base_check_imputation(
+        impute_num_adata, adata_imputed[:, 1:]
+    )
+
+
+def test_base_check_imputation_nan_detected_after_complete_imputation(impute_num_adata):
+    adata_imputed = knn_impute(impute_num_adata, copy=True)
+    adata_imputed.X[0, 2] = np.nan
+    assert not _base_check_imputation(impute_num_adata, adata_imputed)
+
+
+def test_base_check_imputation_nan_detected_after_partial_imputation(impute_num_adata):
+    var_names = ("col2", "col3")
+    adata_imputed = knn_impute(impute_num_adata, var_names=var_names, copy=True)
+    adata_imputed.X[0, 2] = np.nan
+    assert not _base_check_imputation(impute_num_adata, adata_imputed, imputed_var_names=var_names)
+
+
+def test_base_check_imputation_nan_ignored_if_not_in_imputed_column(impute_num_adata):
+    var_names = ("col2", "col3")
+    adata_imputed = knn_impute(impute_num_adata, var_names=var_names, copy=True)
+    # col1 has a NaN at row 2, should get ignored
+    assert _base_check_imputation(impute_num_adata, adata_imputed, imputed_var_names=var_names)
+
+
+def test_base_check_imputation_change_detected_in_non_imputed_column(impute_num_adata):
+    var_names = ("col2", "col3")
+    adata_imputed = knn_impute(impute_num_adata, var_names=var_names, copy=True)
+    # col1 has a NaN at row 2, let's simulate it has been imputed by mistake
+    adata_imputed.X[2, 0] = 42.0
+    assert not _base_check_imputation(impute_num_adata, adata_imputed, imputed_var_names=var_names)
+
+
+def test_base_check_imputation_change_detected_in_imputed_column(impute_num_adata):
+    adata_imputed = knn_impute(impute_num_adata, copy=True)
+    # col3 didn't have a NaN at row 1, let's simulate it has been modified by mistake
+    adata_imputed.X[1, 2] = 42.0
+    assert not _base_check_imputation(impute_num_adata, adata_imputed)
+
+
+# ----------------------------------------------------------------------------------------------------------------------
+
+
+# Base check for imputation (imputation doesn't leave NaN behind, it doesn't affect non-imputated columns, it doesn't
+# affect data in imputated columns that wasn't NaN)
+def _base_check_imputation(
```

Comment:
I'd consider moving this function to the very top. I always try to have fixtures and other test util functions etc at the top so that people see early how the test is set up.